crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,381 @@
+                    }
+                    // If the action is "request", bundle the name in the `RequestAction`, else
+                    // discard the name.
+                    Some(act) => match act.as_ref() {
+                        "request" => RequestAction::Request(match name {
+                            Some(secret_name) => SecretName::from(secret_name),
+                            None => {
+                                return Err(de::Error::missing_field("name"));
+                            }
+                        }),
+                        "request_cancellation" => RequestAction::RequestCancellation,
+                        other_action => RequestAction::_Custom(other_action.to_owned()),
+                    },
+                };
+
+                let requesting_device_id: DeviceIdBox = match requesting_device_id {
+                    None => {
+                        return Err(de::Error::missing_field("requesting_device_id"));
+                    }
+                    Some(id) => id.into(),
+                };
+
+                let request_id = match request_id {
+                    None => {
+                        return Err(de::Error::missing_field("request_id"));
+                    }
+                    Some(id) => id,
+                };
+
+                Ok(RequestToDeviceEventContent::new(
+                    request_action,
+                    requesting_device_id,
+                    request_id,
+                ))
+            }
+        }
+
+        const FIELDS: &[&str] = &["name", "action", "requesting_device_id", "request_id"];
+        deserializer.deserialize_struct(
+            "RequestToDeviceEventContent",
+            FIELDS,
+            RequestToDeviceEventContentVisitor,
+        )
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
+                st.serialize_entry("name", name)?;
+                st.serialize_entry("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                let mut st = serializer.serialize_map(Some(1))?;
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
+        }
+    }
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, StringEnum)]
```

Comment:
Why doesn't this enum have all the goodness the `EventEncryptionAlgorithm` for example has? `PartialEq` and `AsRefStr` would be very useful for this.

Reply:
I've added `PartialEq, Eq, PartialOrd, Ord` to both `SecretName` and `RequestAction`.
`AsRefStr, DisplayAsRefStr, FromString` are already derived through `StringEnum`.